cmd/convertor/builder/builder.go

```diff
@@ -0,0 +1,180 @@
+			return authSplit[0], authSplit[1], nil
+		},
+		PlainHTTP: opt.PlainHTTP,
+	})
+	engineBase, err := getBuilderEngineBase(ctx, resolver, opt.Ref, opt.TargetRef)
+	if err != nil {
+		return nil, err
+	}
+	engineBase.workDir = opt.WorkDir
+	engineBase.oci = opt.OCI
+	var engine builderEngine
+	switch opt.Engine {
+	case BuilderEngineTypeOverlayBD:
+		engine = NewOverlayBDBuilderEngine(engineBase)
+	case BuilderEngineTypeFastOCI:
+		engine = NewFastOCIBuilderEngine(engineBase)
+	}
+	return &overlaybdBuilder{
+		layers: len(engineBase.manifest.Layers),
+		engine: engine,
+	}, nil
+}
+
+func (b *overlaybdBuilder) Build(ctx context.Context) error {
+	defer b.engine.Cleanup()
+	downloaded := make([]chan error, b.layers)
+	converted := make([]chan error, b.layers)
+	var uploaded sync.WaitGroup
+
+	errCh := make(chan error)
+	defer close(errCh)
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// collect error and kill all builder goroutines
+	var retErr error
+	retErr = nil
+	go func() {
+		select {
+		case <-ctx.Done():
+		case retErr = <-errCh:
+		}
+		if retErr != nil {
+			cancel()
+		}
+	}()
+
+	for i := 0; i < b.layers; i++ {
+		downloaded[i] = make(chan error)
+		converted[i] = make(chan error)
+
+		// download goroutine
+		go func(idx int) {
+			defer close(downloaded[idx])
+			if err := b.engine.DownloadLayer(ctx, idx); err != nil {
+				contextSendChan(ctx, errCh, errors.Wrapf(err, "failed to download layer %d", idx))
+				return
+			}
+			logrus.Infof("downloaded layer %d", idx)
+			contextSendChan(ctx, downloaded[idx], nil)
+		}(i)
+
+		// convert goroutine
+		go func(idx int) {
+			defer close(converted[idx])
+			if contextReceiveChan(ctx, downloaded[idx]); ctx.Err() != nil {
+				return
+			}
+			if idx > 0 {
+				if contextReceiveChan(ctx, converted[idx-1]); ctx.Err() != nil {
+					return
+				}
+			}
+			if err := b.engine.BuildLayer(ctx, idx); err != nil {
+				contextSendChan(ctx, errCh, errors.Wrapf(err, "failed to convert layer %d", idx))
+				return
+			}
+			logrus.Infof("layer %d converted", idx)
+			// send to upload(idx) and convert(idx+1) once each
+			contextSendChan(ctx, converted[idx], nil)
+			if idx+1 < b.layers {
+				contextSendChan(ctx, converted[idx], nil)
+			}
+		}(i)
+
+		// upload goroutine
+		uploaded.Add(1)
+		go func(idx int) {
+			defer uploaded.Done()
+			if contextReceiveChan(ctx, converted[idx]); ctx.Err() != nil {
+				return
+			}
+			if err := b.engine.UploadLayer(ctx, idx); err != nil {
+				contextSendChan(ctx, errCh, errors.Wrapf(err, "failed to upload layer %d", idx))
+				return
+			}
+			logrus.Infof("layer %d uploaded", idx)
+		}(i)
+	}
+	uploaded.Wait()
+	if retErr != nil {
+		return retErr
+	}
+
+	if err := b.engine.UploadImage(ctx); err != nil {
+		return errors.Wrap(err, "failed to upload manifest or config")
+	}
+	logrus.Info("convert finished")
+	return nil
+}
+
+// block until ctx.Done() or sended
+func contextSendChan(ctx context.Context, ch chan<- error, value error) {
```

Comment:
contextSendChan and contextReceiveChan only apply to waiting for an error or sending an. A clearer name might be something like GetErrorOnChannel? or SendErrorToChannel?. Tbh not super convinced about those myself but just a suggestion.